piet-direct2d/src/d2d.rs

```diff
@@ -0,0 +1,638 @@
+    pub fn new() -> Result<D2DFactory, Error> {
+        unsafe {
+            let mut ptr: *mut ID2D1Factory1 = null_mut();
+            let hr = D2D1CreateFactory(
+                D2D1_FACTORY_TYPE_MULTI_THREADED,
+                &ID2D1Factory1::uuidof(),
+                &D2D1_FACTORY_OPTIONS {
+                    debugLevel: D2D1_DEBUG_LEVEL_WARNING,
+                },
+                &mut ptr as *mut _ as *mut _,
+            );
+            wrap(hr, ptr, D2DFactory)
+        }
+    }
+
+    // Would it be safe to take &ComPtr<IDXGIDevice> here?
+    pub unsafe fn create_device(&self, dxgi_device: *mut IDXGIDevice) -> Result<D2DDevice, Error> {
+        let mut ptr = null_mut();
+        let hr = self.0.CreateDevice(dxgi_device, &mut ptr);
+        wrap(hr, ptr, D2DDevice)
+    }
+
+    pub unsafe fn get_raw(&self) -> *mut ID2D1Factory1 {
+        self.0.as_raw()
+    }
+
+    pub fn create_path_geometry(&self) -> Result<PathGeometry, Error> {
+        unsafe {
+            let mut ptr = null_mut();
+            let hr = self.0.deref().deref().CreatePathGeometry(&mut ptr);
+            wrap(hr, ptr, PathGeometry)
+        }
+    }
+
+    pub fn create_stroke_style(
+        &self,
+        props: &D2D1_STROKE_STYLE_PROPERTIES,
+        dashes: Option<&[f32]>,
+    ) -> Result<StrokeStyle, Error> {
+        unsafe {
+            let mut ptr = null_mut();
+            let dashes_len = dashes.map(|d| d.len()).unwrap_or(0);
+            assert!(dashes_len <= 0xffff_ffff);
+            let hr = self.0.deref().deref().CreateStrokeStyle(
+                props,
+                dashes.map(|d| d.as_ptr()).unwrap_or(null()),
+                dashes_len as u32,
+                &mut ptr,
+            );
+            wrap(hr, ptr, StrokeStyle)
+        }
+    }
+}
+
+impl D2DDevice {
+    /// Create a new device context from the device.
+    ///
+    /// This is a wrapper for
+    /// [ID2D1Device::CreateDeviceContext](https://docs.microsoft.com/en-us/windows/win32/api/d2d1_1/nf-d2d1_1-id2d1device-createdevicecontext).
+    pub fn create_device_context(&mut self) -> Result<DeviceContext, Error> {
+        unsafe {
+            let mut ptr = null_mut();
+            let options = D2D1_DEVICE_CONTEXT_OPTIONS_NONE;
+            let hr = self.0.CreateDeviceContext(options, &mut ptr);
+            wrap(hr, ptr, DeviceContext)
+        }
+    }
+}
+
+const IDENTITY_MATRIX_3X2_F: D2D1_MATRIX_3X2_F = D2D1_MATRIX_3X2_F {
+    matrix: [[1.0, 0.0], [0.0, 1.0], [0.0, 0.0]],
+};
+
+const DEFAULT_BRUSH_PROPERTIES: D2D1_BRUSH_PROPERTIES = D2D1_BRUSH_PROPERTIES {
+    opacity: 1.0,
+    transform: IDENTITY_MATRIX_3X2_F,
+};
+
+impl DeviceContext {
+    /// Create a new device context from an existing COM object.
+    ///
+    /// Marked as unsafe because the device must be in a good state.
+    /// This *might* be overly conservative.
+    pub unsafe fn new(ptr: ComPtr<ID2D1DeviceContext>) -> DeviceContext {
+        DeviceContext(ptr)
+    }
+
+    /// docs
```

Comment:
"Get the raw pointer."

I don't think this is actually unsafe. Of course, the caller wouldn't be calling it unless they wanted to do unsafe things with it, but this is a fairly clear distinction in Rust.